llvm/lib/CodeGen/ScheduleDAGInstrs.cpp

```diff
@@ -596,6 +596,10 @@ std::optional<unsigned> ScheduleDAGInstrs::initSUnit(MachineInstr &MI) {
   if (MI.isDebugOrPseudoInstr())
     return {};
 
+  // Ensure all instructions can be given a SUnit without re-allocation.
```

Comment:
I was thinking if we could do this in `ScheduleDAGInstrs::enterRegion` to remove this fixed cost from here. What do you think?

Reply:
That would also be fine, the reason I kept it here is to avoid distributing the logic about creating `SUnits` in too many places. But I'm open to discussion :)

Reply:
@martien-de-jong Would you find this better if I do the initialization in `ScheduleDAGInstrs::enterRegion` then?

Reply:
I guess so. I was happy that initSUnit didn't need to care. At the time I hadn't thought about adding SUnits after edge creation. If we can pre-reserve the total number of sunits with some interface I'm fine.